[CI] IncrementalBulkIT testIncrementalBulkHighWatermarkBackOff failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #4272 / rocky-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/dd7xm2xkxppeo)
- [elasticsearch-pull-request #34642 / part-1](https://gradle-enterprise.elastic.co/s/3silzgxkrqad4)
- [elasticsearch-periodic-platform-support #4268 / almalinux-8-aarch64_checkpart1_platform-support-arm](https://gradle-enterprise.elastic.co/s/xyzsgcdtfhgsc)
- [elasticsearch-periodic #4290 / concurrent-search-tests](https://gradle-enterprise.elastic.co/s/g5uajs5q5uhvw)
- [elasticsearch-periodic-platform-support #4260 / oraclelinux-7_platform-support-unix](https://gradle-enterprise.elastic.co/s/tvkqzazr6hsgy)
- [elasticsearch-periodic-platform-support #4256 / windows-2016_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/zjd7y3dtk3udm)

**Reproduction Line:**
```
./gradlew ":server:internalClusterTest" --tests "org.elasticsearch.action.bulk.IncrementalBulkIT.testIncrementalBulkHighWatermarkBackOff" -Dtests.seed=CFF58A9BBC3643B -Dtests.locale=ar-PS -Dtests.timezone=America/Tortola -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testIncrementalBulkHighWatermarkBackOff'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.action.bulk.IncrementalBulkIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: Unexpected ShardFailures: failure in bulk execution:
[0]: index [test], id [8F0rWZIBT0LgtC87Kmxj], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
[1]: index [test], id [8V0rWZIBT0LgtC87Kmxj], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
[2]: index [test], id [8l0rWZIBT0LgtC87Kmxj], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
[3]: index [test], id [810rWZIBT0LgtC87Kmxj], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
[4]: index [test], id [9F0rWZIBT0LgtC87Kmxj], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
[5]: index [test], id [9V0rWZIBT0LgtC87Kmxj], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
[6]: index [test], id [null], message [java.lang.IllegalArgumentException: autoGeneratedTimestamp should not be set externally]
Expected: is <false>
     but: was <true>
```

**Issue Reasons:**
- [main] 6 failures in test testIncrementalBulkHighWatermarkBackOff (2.2% fail rate in 273 executions)
- [main] 4 failures in pipeline elasticsearch-periodic-platform-support (80.0% fail rate in 5 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/server/src/internalClusterTest/java/org/elasticsearch/action/bulk/IncrementalBulkIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/action/bulk/IncrementalBulkIT.java
@@ -233,6 +233,9 @@ public class IncrementalBulkIT extends ESIntegTestCase {
 
         handlers.add(handlerThrottled);
 
+        // Wait until we are ready for the next page
+        assertBusy(() -> assertTrue(nextPage.get()));
+
         for (IncrementalBulkService.Handler h : handlers) {
             refCounted.incRef();
             PlainActionFuture<BulkResponse> future = new PlainActionFuture<>();

--- a/server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java
+++ b/server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java
@@ -105,6 +105,7 @@ public class IncrementalBulkService {
         private boolean closed = false;
         private boolean globalFailure = false;
         private boolean incrementalRequestSubmitted = false;
+        private boolean bulkInProgress = false;
         private ThreadContext.StoredContext requestContext;
         private Exception bulkActionLevelFailure = null;
         private long currentBulkSize = 0L;
@@ -130,6 +131,7 @@ public class IncrementalBulkService {
 
         public void addItems(List<DocWriteRequest<?>> items, Releasable releasable, Runnable nextItems) {
             assert closed == false;
+            assert bulkInProgress == false;
             if (bulkActionLevelFailure != null) {
                 shortCircuitDueToTopLevelFailure(items, releasable);
                 nextItems.run();
@@ -143,6 +145,7 @@ public class IncrementalBulkService {
                             requestContext.restore();
                             final ArrayList<Releasable> toRelease = new ArrayList<>(releasables);
                             releasables.clear();
+                            bulkInProgress = true;
                             client.bulk(bulkRequest, ActionListener.runAfter(new ActionListener<>() {
 
                                 @Override
@@ -158,6 +161,7 @@ public class IncrementalBulkService {
                                     handleBulkFailure(isFirstRequest, e);
                                 }
                             }, () -> {
+                                bulkInProgress = false;
                                 requestContext = threadContext.newStoredContext();
                                 toRelease.forEach(Releasable::close);
                                 nextItems.run();
@@ -177,6 +181,7 @@ public class IncrementalBulkService {
         }
 
         public void lastItems(List<DocWriteRequest<?>> items, Releasable releasable, ActionListener<BulkResponse> listener) {
+            assert bulkInProgress == false;
             if (bulkActionLevelFailure != null) {
                 shortCircuitDueToTopLevelFailure(items, releasable);
                 errorResponse(listener);
@@ -187,7 +192,9 @@ public class IncrementalBulkService {
                         requestContext.restore();
                         final ArrayList<Releasable> toRelease = new ArrayList<>(releasables);
                         releasables.clear();
-                        client.bulk(bulkRequest, ActionListener.runBefore(new ActionListener<>() {
+                        // We do not need to set this back to false as this will be the last request.
+                        bulkInProgress = true;
+                        client.bulk(bulkRequest, ActionListener.runAfter(new ActionListener<>() {
 
                             private final boolean isFirstRequest = incrementalRequestSubmitted == false;
 

